lib/src/str_util.rs

```diff
@@ -32,105 +32,189 @@ pub enum StringPatternParseError {
+    /// `exact[-i]:`, `substring[-i]`, `glob[-i]:`, or or `i:` (short for
+    /// `{default_kind}-i:`), a pattern of the specified kind is returned.
+    /// Returns an error if the string has an unrecognized prefix. Otherwise,
+    /// a pattern of kind `default_kind` is returned.
+    pub fn parse_with_default_kind(
+        src: &str,
+        default_kind: &str,
+    ) -> Result<Self, StringPatternParseError> {
+        let (maybe_kind, pat) = match src.split_once(':') {
+            Some((kind, pat)) => (Some(kind), pat),
+            None => (None, src),
+        };
+        Self::from_str_maybe_kind(pat, maybe_kind, default_kind)
     }
 
-    /// Parses the given string as a `StringPattern`. Everything before the
-    /// first ":" is considered the string's prefix. If the prefix is "exact:",
-    /// "glob:", or "substring:", a pattern of the specified kind is returned.
-    /// Returns an error if the string has an unrecognized prefix. Otherwise, a
-    /// `StringPattern::Exact` is returned.
+    /// Helper for `StringPattern::parse_with_default_kind(src, "exact')`. See
+    /// [`StringPattern::parse_with_default_kind()`] for details.
     pub fn parse(src: &str) -> Result<StringPattern, StringPatternParseError> {
-        if let Some((kind, pat)) = src.split_once(':') {
-            StringPattern::from_str_kind(pat, kind)
-        } else {
-            Ok(StringPattern::exact(src))
-        }
+        Self::parse_with_default_kind(src, "exact")
     }
 
     /// Creates pattern that matches exactly.
-    pub fn exact(src: impl Into<String>) -> Self {
-        StringPattern::Exact(src.into())
+    pub fn exact(src: impl Into<String>, case_sensitivity: CaseSensitivity) -> Self {
+        StringPattern::Exact(src.into(), case_sensitivity)
     }
 
     /// Parses the given string as glob pattern.
-    pub fn glob(src: &str) -> Result<Self, StringPatternParseError> {
+    pub fn glob(
+        src: &str,
+        case_sensitivity: CaseSensitivity,
+    ) -> Result<Self, StringPatternParseError> {
         // TODO: might be better to do parsing and compilation separately since
         // not all backends would use the compiled pattern object.
         // TODO: if no meta character found, it can be mapped to Exact.
         let pattern = glob::Pattern::new(src).map_err(StringPatternParseError::GlobPattern)?;
-        Ok(StringPattern::Glob(pattern))
+        Ok(StringPattern::Glob(pattern, case_sensitivity))
     }
 
-    /// Parses the given string as pattern of the specified `kind`.
+    /// Parses the given string as a pattern of the specified `kind`.
     pub fn from_str_kind(src: &str, kind: &str) -> Result<Self, StringPatternParseError> {
-        match kind {
-            "exact" => Ok(StringPattern::exact(src)),
-            "glob" => StringPattern::glob(src),
-            "substring" => Ok(StringPattern::Substring(src.to_owned())),
+        let (base_kind, case_sensitivity) = match kind.strip_suffix("-i") {
```

Comment:
nit: I'll list "exact", "exact-i", ... without resolving `case_sensitivity` separately. If we add "regex:", "regex:" and "regex-i:" will be processed differently.

Reply:
I think it could be similar; we’d do e.g. `RegexBuilder::new(pat).case_insensitive(case_sensitivity == CaseSensitivity::Insensitive).build()`. But admittedly that’s not exactly saving much verbosity at that point and I probably got too clever here; I can change it.